File: src/Lista03/lab3.ml
let nComp f n = fun x ->
    if n <= 0 then []
    else
        let rec helper elem i =
            match i with
                0 -> []
                |_ -> elem::(helper (f elem) (i-1))
        in helper x n
    ;;

nComp (fun x -> x * x) 4 3 = [3; 9; 81; 6561];;
nComp (fun x -> x / 2) (-1) 8 = [];;
nComp (fun x -> x * x * x) 3 (-2) = [-2; -8; -512];;

let area (a, b) f n =
    let rec sieve from_n =
        if from_n <= n then from_n::sieve(from_n + 1) else [] in
    let interval_len = (b -. a) /. float(n)
    in List.fold_left (+.) 0. (List.map (fun x -> abs_float(f(a +. interval_len *. float(x))) *. interval_len) (sieve 1))
;;

let f (x: float): float = x *. x;;
let f1 (x: float): float = 4. -. x;;

area (0., 3.) f 100 = 9.13544999999999874;;
area (1.0, 2.0) f 1500 = 2.33433340740740913;;
area (0., 4.) f1 100000 = 7.99992000000000125;;
area (1., 2.) f1 2000 = 2.49975;

File: src/Lista03/lab3.scala
def nComp[A](f: A => A)(n: Int)(x: A): List[A] =
    if n <= 0 then Nil
    else
        def helper(elem: A, i: Int): List[A] =
            if i == 0 then Nil
            else elem :: helper(f(elem), i - 1)

        helper(x, n)

nComp((x: Int) => x * x)(4)(3) == List(3, 9, 81, 6561)
nComp((x: Int) => x / 2)(-1)(8) == List()
nComp((x: Int) => x * x * x)(3)(-2) == List(-2, -8, -512)


def area(a: Double, b: Double)(f: Double => Double)(n: Int): Double =
    val xs = List.range(1, n)
    val interval_len = (b - a) / n
    val areas = xs map (x => (f(a + interval_len * x)).abs * interval_len)
    (areas foldLeft 0.0)((acc, x) => acc + x)


area(0.0, 3.0)(x => x * x)(100) == 8.86545
area(1.0, 2.0)(x => x * x)(1500) == 2.3316667407407423
area(0.0, 4.0)(x => 4 - x)(100000) == 7.999920000000001
area(1.0, 2.0)(x => 4 - x)(2000) == 2.4987500000000002